### PR TITLE
Refactor Test Definition and add capability of test with flexible SetupConfig/Deployment Parameters for selected test cases when testing against Azure platform

### PR DESCRIPTION
**Pull request by harz566**
- Refactor Test Definition and wrap everything about deployment/provision into `<SetupConfig>`, to replace previous `<AdditionalHWConfig>`, because Azure Deployment may not related to 'Hardware', some of those actual configurations are 'Software' configurations.

- Enable Test flexibility with configurable and Comma-Separated multi values in '-ARMImageName', '-TestLocation', '-OverrideVMSize', '-OsVHD'. All previous parameters can be used as usual, in addition, introduce  another new parameter '-ForceCustom' to force override/customize the test setup parameters for deployment, even when test cases has pre-defined configurations. See the updated README.md

- Test Case can define their expected Setup environment by design, with additional `<TestLocation>` , `<ARMImageName>` , `<OsVHD>` as child elements of `<SetupConfig>` from test case definition xml.  And inner-text content of those xml elements can also be a Comma-Separated value, which means testing against multiple setup environments, just like the parameter value provided from Run-LISAv2 command line

- Removing some $global:XXX variables to keep clean in source code from abuse of global variables, such as '$global:ARMImageName', '$global:VMGeneration', '$global:TestLocation', '$global.OverrideVMSize'.

- Simplified parameters to Run-LISAv2

> - '-TestLocation' is optional when testing with Azure platform, LISAv2 will automatically select one enabled region to do testing, see detailed explain from README.md

> - '-ARMImageName' is optional when testing with Azure platform, (but user can still use it as before, still take affect). When '-ARMImageName' and '-OsVHD' both are not provided, LISAv2 will pick the `<DefaultARMImageName>` from .\XML\GlobalConfiguration.xml.  This is also how we can handle most of test cases does not have a preferred distro to be tested with, but special/certain Test Case can define their expected `<ARMImageName>` from its' `<SetupConfig>` definition.

> - '-ForceCustom' is introduced here as a new parameter of LISAv2.  It's optional when testing on Azure platform, but it provides a clear and explicit meaning for users to Run-LISAv2 with more flexible configurations when provisioning on Azure for Test Cases. At the meantime, '-ForceCustom' provides the capability to achieve the same impact when user test with '-OverrideVMSize' parameter before this PR.   Which means when user wants to override in force for the pre-defined `<ARMImageName>, <TestLocation>, <OverrideVMSize>, <OsVHD>, <Networking>, <OSType>,<DiskType>,<ImageType>,<...>` values for each Test Case, the '-ForceCustom' switch will give LISAv2 a clear indication of always using parameter values provided from Run-LISAv2 command line to do test environment deployment and provision on Azure, and **NOT** respect TestCase defined SetupConfigurations

- Updated 'Ready' platform parameters to use '-TestLocation' instead of previous '-RGIdentifier'. TestLocation of 'Ready' platform should follow this Comma-Separated format: `<public_ip_address_a>:<ssh_port_a>,<public_ip_address_b>:<ssh_port_b>`   For example:  `10.100.100.100:1111,10.100.100.100:1112`
 
======Run-LISAv2  command **without '-ARMImageName'** and  **no '-TestLocation'**================
. 'C:\LISAv2\Run-LisaV2.ps1' -TestPlatform 'Azure' -CustomTestParameters 'DPDK_SOURCE_URL=https://git.dpdk.org/dpdk/snapshot/dpdk-v20.02.tar.gz' -XMLSecretFile 'C:\LISAv2\AzureSecretsTestOnly.xml' -RGIdentifier harz0702 -TestNames 'VERIFY-DEPLOYMENT-PROVISION,VALIDATE-INTEL-SGX-DRIVER-FOR-DC-VM,KDUMP-CRASH-AUTO-SIZE,VERIFY-DPDK-COMPLIANCE'

=======Test Logs/results with **different ARMImageName**, and **different TestLocation** ====================
[LISAv2 Test Results Summary]
Test Run On           : 07/06/2020 00:03:14
Initial Kernel Version: 5.3.0-1028-azure
Final Kernel Version  : 5.3.0-1028-azure
Total Test Cases      : 4 (4 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:26

   ID TestArea             TestCaseName-SetupHashCode                                                  TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 KDUMP                KDUMP-CRASH-AUTO-SIZE                                                             PASS                  4.6 
      ARMImageName: Redhat RHEL 7.7 Latest, SetupType: OneVM, TestLocation: westus2, VMGeneration: 1
    2 DPDK                 VERIFY-DPDK-COMPLIANCE                                                            PASS                 5.95 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, Networking: SRIOV, SetupType: OneVM3NIC, TestLocation: westus2, VMGeneration: 1
    3 SGX                  VALIDATE-INTEL-SGX-DRIVER-FOR-DC-VM                                               PASS                  4.1 
      ARMImageName: Canonical UbuntuServer 18_04-lts-gen2 latest, OverrideVMSize: Standard_DC2s_v2, SetupType: OneVM, TestLocation: uksouth, VMGeneration: 2
    4 CORE                 VERIFY-DEPLOYMENT-PROVISION                                                       PASS                 2.06 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, SetupType: OneVM, VMGeneration: 1, TestLocation: westus2
      FirstBoot : PASS
      FirstBoot : Call Trace Verification : PASS
      Reboot : PASS
      Reboot : Call Trace Verification : PASS
      

Logs can be found at C:\LISAv2\TestResults\2020-07-05-17-02-13-7744

07/06/2020 00:29:30 : [DEBUG] Creating 'C:\LISAv2\Azure-IV58-TestLogs.zip' from 'C:\LISAv2\TestResults\2020-07-05-17-02-13-7744'
07/06/2020 00:29:30 : [DEBUG] C:\LISAv2\Azure-IV58-TestLogs.zip created successfully.
07/06/2020 00:29:30 : [INFO ] Analyzing test results ...
07/06/2020 00:29:30 : [INFO ] LISAv2 exit code: 0

=========Another Test ===================
. 'C:\LISAv2\Run-LisaV2.ps1' -TestPlatform 'Azure' -RGIdentifier 'harz-dpdk' -ARMImageName '**Canonical UbuntuServer 18.04-LTS Latest,Redhat RHEL 7.7 Latest**' -OverrideVMSize '**Standard_DS1_v2,Standard_DS2_v2**' -CustomTestParameters 'DPDK_SOURCE_URL=http://git.dpdk.org/dpdk-stable/snapshot/dpdk-stable-18.11.9-rc2.tar.gz' -XMLSecretFile 'C:\LISAv2\AzureSecretsTestOnly.xml' -TestNames 'WALA-PROCESS-CHECK,WALA-VERIFY-MNT-RESOURCE-WRITABLE' -EnableTelemetry **-TestIterations 2**

07/13/2020 19:28:15 : [INFO ] Test OF97 finished
07/13/2020 19:28:15 : [INFO ] 
[LISAv2 Test Results Summary]
Test Run On           : 07/13/2020 19:09:39
**ARM Image Under Test  : Canonical : UbuntuServer : 18.04-LTS : Latest
ARM Image Under Test  : Redhat : RHEL : 7.7 : Latest
Override VM size Under Test  : Standard_DS1_v2
Override VM size Under Test  : Standard_DS2_v2**
Initial Kernel Version: 3.10.0-1062.el7.x86_64
Final Kernel Version  : 3.10.0-1062.el7.x86_64
Total Test Cases      : 16 (16 Passed, 0 Failed, 0 Aborted, 0 Skipped)
Total Time (dd:hh:mm) : 0:0:18

   ID TestArea             TestCaseName                                                                TestResult TestDuration(in minutes) 
-------------------------------------------------------------------------------------------------------------------------------------------
    1 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.63 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
    2 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.64 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
    3 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.46 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2
    4 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.58 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2
    5 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.62 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
    6 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.53 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
    7 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.53 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2
    8 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.53 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS1_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2
    9 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.63 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
   10 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.61 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
   11 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.61 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2
   12 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.61 
      ARMImageName: Canonical UbuntuServer 18.04-LTS Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2
   13 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.55 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
   14 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.55 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 1, TestLocation: westus2
   15 WALA                 WALA-PROCESS-CHECK                                                                PASS                 0.54 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2
   16 WALA                 WALA-VERIFY-MNT-RESOURCE-WRITABLE                                                 PASS                 0.55 
      ARMImageName: Redhat RHEL 7.7 Latest, OverrideVMSize: Standard_DS2_v2, SetupType: OneVM, TestIteration: 2, TestLocation: westus2

Logs can be found at C:\LISAv2\TestResults\2020-07-13-12-08-49-9382

07/13/2020 19:28:15 : [DEBUG] Creating 'C:\LISAv2\Azure-OF97-TestLogs.zip' from 'C:\LISAv2\TestResults\2020-07-13-12-08-49-9382'
07/13/2020 19:28:15 : [DEBUG] C:\LISAv2\Azure-OF97-TestLogs.zip created successfully.
07/13/2020 19:28:15 : [INFO ] Analyzing test results ...
07/13/2020 19:28:15 : [INFO ] LISAv2 exit code: 0

### CODE REVIEW

File: LISAv2-Framework.psm1 [hunk 1]
```diff
@@ -72,7 +72,8 @@ function Start-LISAv2 {
 		[string] $ResultDBTable = "",
 		[string] $ResultDBTestTag = "",
 
-		[switch] $ExitWithZero
+		[switch] $ExitWithZero,
+		[switch] $ForceCustom
```
Comment (LiliDeng) [inline LISAv2-Framework.psm1:76]: where did you use it?

File: Libraries/Azure.psm1 [hunk 2]
```diff
@@ -2100,7 +2120,7 @@ Function Get-AllDeploymentData([string]$ResourceGroups, [string]$PatternOfResour
 		Add-Member -InputObject $objNode -MemberType NoteProperty -Name URLv6 -Value $URL -Force
 		Add-Member -InputObject $objNode -MemberType NoteProperty -Name Status -Value $Status -Force
 		Add-Member -InputObject $objNode -MemberType NoteProperty -Name InstanceSize -Value $InstanceSize -Force
-		Add-Member -InputObject $objNode -MemberType NoteProperty -Name VMGeneration -Value $VMGeneration -Force
+		#Add-Member -InputObject $objNode -MemberType NoteProperty -Name VMGeneration -Value $VMGeneration -Force
```
Comment (LiliDeng) [inline Libraries/Azure.psm1]: when we use osvhd, we need specify the generation by the user

File: LISAv2-Framework.psm1 (hunk 1 above)
Comment (harz566) [reply LISAv2-Framework.psm1:76]: CommonFunctions.psm1, line 256 : Add-SetupConfig

File: Libraries/Azure.psm1 (hunk 2 above)
Comment (harz566) [reply Libraries/Azure.psm1]: As of 'specify the generation of VM/OsVHD', That's already handled in Azure Controller and HyperV Controller.  I searched out, the VMData object does not need the $VMGeneration property so far.

**Comment by juhlee-microsoft** [comment]
@harz566 Any chance you find the timeline when it can be merged? It's been open long

**Comment by harz566** [comment]
> @harz566 Any chance you find the timeline when it can be merged? It's been open long

Hi @juhlee-microsoft , @LiliDeng and I are proactively testing this tremendous change, hopefully it will be merged in coming one week.

**Review by LiliDeng — DISMISSED** [review]

**Review by LiliDeng — APPROVED** [review]

**Pull request merged**